### PR TITLE
Weapon Holsters

### PR DESCRIPTION
**Pull request by 97saundersj**
### Added shoulder Holsters for Weapon Switching

When enabled, you must hover your main hand over a shoulder or upper back to activate a holster and press the switch weapon binding to change weapons. These changes also work with the left handed mode.

The files are available here for testing: [HaloCEVR.zip](https://github.com/user-attachments/files/18035230/HaloCEVR.zip)

The holsters are configurable with the following config settings:
- EnableWeaponHolsters: [Bool]

- LeftShoulderHolsterDistance: [Float]
- LeftShoulderHolsterOffset: [Vector3]
- RightShoulderHolsterDistance: [Float]
- RightShoulderHolsterOffset: [Vector3]

The default values work well for me, but others may need to adjust the offset and distance settings.

To calculate the holster locations, I offset based on the HMD location. This works well most of the time, but if the user has their head turned significantly to one side, the holster may not activate correctly over the far edges of the shoulders. To try and counter this I have made a large activation distance by default, If anyone has suggestions on how to better calculate the holster locations, please let me know!

### CODE REVIEW

File: HaloCEVR/Game.cpp [hunk 1]
```diff
@@ -787,6 +787,10 @@ void Game::SetupConfigs()
 	c_ShowRoomCentre = config.RegisterBool("ShowRoomCentre", "Draw an indicator at your feet to show where the player character is actually positioned", true);
 	c_d3d9Path = config.RegisterString("CustomD3D9Path", "If set first try to load d3d9.dll from the specified path instead of from system32", "");
 
+	// Experimental settings
+	c_LeftHandShoulderHolsterDistance = config.RegisterFloat("LeftHandShoulderHolster", "Experimental - Bringing the left hand within this distance of a shoulder will swap weapons (<0 to disable)", 0.3f);
+	c_RightHandShoulderHolsterDistance = config.RegisterFloat("RightHandShoulderHolster", "Experimental - Bringing the right hand within this distance of a shoulder will swap weapons (<0 to disable)", 0.3f);
+
```
[the remark this answers is not in the document]
Comment (97saundersj) [reply HaloCEVR/Game.cpp:794]: Thanks for the feedback, I have defaulted to -1.

**Comment by 97saundersj** [comment]
Fixed an issue raised on discord where the holsters would stop working properly after a short amount of time.

File: HaloCEVR/Game.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (97saundersj) [reply HaloCEVR/Game.cpp:798]: Actually instead I have set the value by default but added a setting to enable / disable weapon holsters which is disabled by default.

**Review by HaMM4R — APPROVED** [review]
Ah yeah my bad nice work :)

**Comment by HoriZonUK** [comment]
Great addon works really well :)

**Comment by teddybear082** [comment]
Great work, this works as expected for me, using Quest 3 and Virtual Desktop.  I even tried while crouching and after having turned around in real life and it still accurately switched the weapon. I also put my hands up in sniping position and it did not trigger the weapon switch. Quick technical question - does this depend on what button is bound to "switch weapon" / is there a new button binding in steam vr for this action or is it always grip no matter what? I'm thinking that now that I have this I might be able to try re-binding the interact and reload button from my hand rest to the grip button since when my hand is out in front of me it won't trigger switch weapon and would only reload or do interact.  But not sure if that would mess up anything in your code.

**Comment by slipperfish** [comment]
Thanks for this feature, it's becoming of halo's two weapon inventory and it feels like how it would be done officially. 

Using the quest 2 here. I'm also a squeezer and casually hold grip buttons. When I rest my palm/wrist of my dominant hand against my chest to simulate a stock, turning my head triggers the weapon switch. Also if I draw a weapon quickly it triggers a melee attack, which is pretty cool but the out of place animation and inability to shoot means this isn't a feature. I'll mess with the hitbox size and melee threshold and report back later.

I like that you're doing this to free up grip buttons. Not sure if this is the right place to give feature requests, but you might be interested. If the two handed weapon hold also had a hitbox, the left grip could also be used for interactions or possibly grenades. And I'd stop pressing a two handed hold accidentally.

**Comment by slipperfish** [comment]
Another thing I thought about is if you add the condition that the controller needs to be in an expected orientation when performing a behind the shoulder grab (trigger up or upside down), then you can make the holster size generous as it should be and not worry about accidental weapon swaps.

**Comment by 97saundersj** [comment]
@slipperfish Thanks for the feedback!
That's a good point about it accidentally switching when you bring your hand close to you whilst holding down the button. I've updated the holster weapon switch to require a button press once your hand is in a holster, so this should fix one of your issues.
With the accidental melee, I could look at maybe preventing a melee if your hand ends up in a holster or something but it might be better to just tweak the config settings like you mentioned. Let me know if that worked for you.
The two handed weapon proximity feature is also something I've been looking at recently so I hope to get those changes ready to go soon :)

File: HaloCEVR/Game.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (HaMM4R) [reply HaloCEVR/Game.cpp:798]: Thinking about it I really don't think this should be off by default, its something that benefits the experience and a lot of users might miss this and not realise its a feature, I can't think of any reason why you wouldn't have it on by default.

[the remark this answers is not in the document]
Comment (97saundersj) [reply HaloCEVR/Game.cpp:798]: I agree with that, but I feel like that's up to the mod owners to decide. If they want to make it enabled by default once it's merged in then I wouldn't have any complaints :)

[the remark this answers is not in the document]
Comment (LivingFray) [reply HaloCEVR/Game.cpp:798]: I'd bet 95% of users won't even look at the config file, so better to default this to on so people actually know it exists.

**Comment by slipperfish** [comment]
When I use the updated zipfile it won't start in VR and hangs at "next up" in Steam VR, with the game running on my desktop with a mirror view. I'm guessing you use that for testing without the headset and forgot to switch it back.

File: HaloCEVR/Game.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (97saundersj) [reply HaloCEVR/Game.cpp:798]: @LivingFray thanks for the reply! I have enabled the weapon holsters by default now and updated the readme to reflect this.

**Comment by 97saundersj** [comment]
@slipperfish You're right, sorry I packaged up the wrong files. Fixed it now.

**Pull request merged**